### PR TITLE
declared-license-mapping: Strip a trailing white space

### PR DESCRIPTION
**Pull request by fviernau**
declared-license-mapping: Strip a trailing white space in a declared license

This entry was originally added in b3595cd for a Maven package, but since
a49dbba we strip declared licenses coming from Maven, so also strip the
lookup key in the mapping.

A future change will ensure that we consistently do trimming of declared
licenses for all package managers.

Signed-off-by: Frank Viernau <frank.viernau@here.com>

### CODE REVIEW

**Comment by sschuberth** [comment]
Shouldn't the commit you refer to in the commit message be b3595cd7bf55f4ef16a20f523853cbfa088599ca, as that already added

         """Apache License, Version 2.0 and
     Common Development And Distribution License (CDDL) Version 1.0 """.trimIndent()
                 to enumSetOf(SpdxLicense.APACHE_2_0, SpdxLicense.CDDL_1_0),

(which contains the trailing space)?

**Comment by sschuberth** [comment]
Also, are you sure that the trailing space actually *is* wrong? I mean, whoever adds newlines to license strings probably also adds trailing spaces...

**Comment by fviernau** [comment]
> Also, are you sure that the trailing space actually _is_ wrong? I mean, whoever adds newlines to license strings probably also adds trailing spaces...

Hm, shouldn't ORT always strip leading / trailing white spaces before performing the declared license map look-up? If so, then leading and trailing white spaces do not make any sense, right?

**Comment by fviernau** [comment]
> Shouldn't the commit you refer to in the commit message be [b3595cd](https://github.com/oss-review-toolkit/ort/commit/b3595cd7bf55f4ef16a20f523853cbfa088599ca), as that already added

My bad, fixed it. Thanks!

**Comment by sschuberth** [comment]
> > Also, are you sure that the trailing space actually _is_ wrong? I mean, whoever adds newlines to license strings probably also adds trailing spaces...
> 
> Hm, shouldn't ORT always strip leading / trailing white spaces before performing the declared license map look-up? If so, then leading and trailing white spaces do not make any sense, right?

But there is no stripping / trimming performed, the YAML is read straight into a map, with unmodified keys:

https://github.com/oss-review-toolkit/ort/blob/9dfdff59c1fda754ee781377c15ba4563c158562/spdx-utils/src/main/kotlin/SpdxDeclaredLicenseMapping.kt#L34-L35

**Comment by fviernau** [comment]
> And I believe it should also stay like this; any "magic" here like trimming keys is intransparent to the editors of `declared-license-mapping.yml`, IMO.

I didn't mean that the values inserted into the map should be trimmed.
I was wondering whether the values which are used as keys to look-up a value should be trimmed.

**Comment by sschuberth** [comment]
> > And I believe it should also stay like this; any "magic" here like trimming keys is intransparent to the editors of `declared-license-mapping.yml`, IMO.
> 
> I didn't mean that the values inserted into the map should be trimmed.
> I was wondering whether the values which are used as keys to look-up a value should be trimmed.

What you call "values which are used as keys" is what I call just "keys" above; and these are not trimmed, and should not be trimmed IMO.

**Comment by fviernau** [comment]
> What you call "values which are used as keys" is what I call just "keys" above; and these are not trimmed, and should not be trimmed IMO. _Edit:_ Ok, now I got what you mean, but please see my sentence below.
> 
> An as a result, also the keys used for lookups should not be trimmed, as otherwise you can never match untrimmed keys in the map.

In my opinion it would be best to always trim to get rid of confusion and to make things simpler.
But I'm fine with disagreeing on this point.

If we do disagree, then I'd file a PR removing this call here: https://github.com/oss-review-toolkit/ort/blob/9dfdff59c1fda754ee781377c15ba4563c158562/analyzer/src/main/kotlin/managers/utils/MavenSupport.kt#L131
Is this what you prefer?

**Comment by sschuberth** [comment]
> If we do disagree, then I'd file a PR removing this call here:

That is important context which I was lacking so far. Your commit message sounded as if wanted to fix an entry which was wrong as it never occurred like this in any original package manager's meta data. Instead, what you wanted to do is to account for our Maven implementation to trim the declared license meta data. With that rationale I would have understood and accepted the change.

But now that you opened this can of worms I briefly looked our package manager implementations, and it seems we're inconsistent about for which package managers we trim the declared license string, and for which we don't.

Basically, I like the idea to capture the original data from the package manager as closely as possible, i.e. without trimming. But sometimes that's not possible anyway as we need to parse a string like "LicenseA / LicenseB" into a list of "LicenseA", "LicenseB" where we already do trimming. So, if we'd consistently do trimming for *all* package managers, I'd also be fine to trim the keys in `declared-license-mapping`, obviously.

TL;DR, could you please add the "account for our Maven implementation to trim the declared license" part to your commit message, so we can merge this and do a follow-up commit that ensures all package managers trim declared license strings?

**Comment by fviernau** [comment]
> That is important context which I was lacking so far. Your commit message sounded as if wanted to fix an entry which was wrong as it never occurred like this in any original package manager's meta data. Instead, what you wanted to do is to account for our Maven implementation to trim the declared license meta data. With that rationale I would have understood and accepted the change.
> 
> But now that you opened this can of worms I briefly looked our package manager implementations, and it seems we're inconsistent about for which package managers we trim the declared license string, and for which we don't.
> 
> Basically, I like the idea to capture the original data from the package manager as closely as possible, i.e. without trimming. But sometimes that's not possible anyway as we need to parse a string like "LicenseA / LicenseB" into a list of "LicenseA", "LicenseB" where we already do trimming. So, if we'd consistently do trimming for _all_ package managers, I'd also be fine to trim the keys in `declared-license-mapping`, obviously.
> 
> TL;DR, could you please add the "account for our Maven implementation to trim the declared license" part to your commit message, so we can merge this and do a follow-up commit that ensures all package managers trim declared license strings?

Agreed.

**Comment by sschuberth** [comment]
Sorry, more commit message nits as I was again looking deeper at the issue: The trimming was only introduced by a49dbba72c6b83779e6767a8eb31286c912e7cae, i.e. in commit b3595cd7bf55f4ef16a20f523853cbfa088599ca it was correct to have the trailing space, as back then no trimming occurred. So the commit message should probably say something like

```
declared-license-mapping: Strip a trailing white space in a declared license

This entry was originally added in b3595cd for a Maven package, but since
a49dbba we strip declared licenses coming from Maven, so also strip the
lookup key in the mapping.

A future change will ensure that we consistently do trimming of declared
licenses for all package managers.
```

**Comment by fviernau** [comment]
> Sorry, more commit message nits as I was again looking deeper at the issue: The trimming was only introduced by [a49dbba](https://github.com/oss-review-toolkit/ort/commit/a49dbba72c6b83779e6767a8eb31286c912e7cae), i.e. in commit [b3595cd](https://github.com/oss-review-toolkit/ort/commit/b3595cd7bf55f4ef16a20f523853cbfa088599ca) it was correct to have the trailing space, as back then no trimming occurred. So the commit message should probably say something like

done.

**Review by sschuberth — APPROVED** [review]
